packages/neo-one-client-core/src/user/DapiUserAccountProvider.ts

```diff
@@ -0,0 +1,394 @@
+  private readonly networksInternal$: BehaviorSubject<ReadonlyArray<NetworkType>>;
+  private readonly defaultNetworkInternal$: BehaviorSubject<NetworkType | undefined>;
+
+  public constructor({ dapi, provider }: { readonly dapi: Dapi; readonly provider: TProvider }) {
+    super({ provider });
+    this.dapi = dapi;
+    this.initPromise = this.init();
+
+    this.userAccountsInternal$ = new BehaviorSubject<ReadonlyArray<UserAccount>>([]);
+    this.currentAccountInternal$ = new BehaviorSubject<UserAccount | undefined>(undefined);
+    this.networksInternal$ = new BehaviorSubject<ReadonlyArray<NetworkType>>([]);
+    this.defaultNetworkInternal$ = new BehaviorSubject<NetworkType | undefined>(undefined);
+
+    let firstCall: typeof ACCOUNT_CHANGED | typeof NETWORK_CHANGED | undefined;
+
+    const updateAccount = ({ address, label }: DapiAccount) =>
+      this.dapi.getPublicKey().then(({ publicKey }) => {
+        const network = this.defaultNetworkInternal$.getValue();
+        if (network !== undefined) {
+          this.updateUserAccount({
+            network,
+            publicKey,
+            address,
+            label,
+          });
+        }
+      });
+
+    const setAccountChangedEventListener = () => {
+      this.dapi.addEventListener(ACCOUNT_CHANGED, (dapiAccount) => {
+        if (firstCall === undefined) {
+          firstCall = ACCOUNT_CHANGED;
+        }
+        if (firstCall === ACCOUNT_CHANGED) {
+          new Promise<void>((resolve) => {
+            this.dapi.addEventListener(NETWORK_CHANGED, (dapiNetworks) => {
+              this.updateNetworks(dapiNetworks);
+              resolve();
+              this.dapi.removeEventListener(NETWORK_CHANGED);
+              setNetworkChangedEventListener();
+            });
+          })
+            .then(() => {
+              updateAccount(dapiAccount).catch(() => {
+                // do nothing
+              });
+            })
+            .catch(() => {
+              // do nothing
+            });
+
+          firstCall = undefined;
+        }
+      });
+    };
+
+    const setNetworkChangedEventListener = () => {
+      this.dapi.addEventListener(NETWORK_CHANGED, (dapiNetworks) => {
+        if (firstCall === undefined) {
+          firstCall = NETWORK_CHANGED;
+        }
+        if (firstCall === NETWORK_CHANGED) {
+          this.updateNetworks(dapiNetworks);
+          new Promise<void>((resolve) =>
+            this.dapi.addEventListener(ACCOUNT_CHANGED, (dapiAccount) => {
+              updateAccount(dapiAccount).catch(() => {
+                // do nothing
+              });
+              resolve();
+              this.dapi.removeEventListener(ACCOUNT_CHANGED);
+              setAccountChangedEventListener();
+            }),
+          ).catch(() => {
+            // do nothing
+          });
+        }
+
+        firstCall = undefined;
+      });
+    };
+
+    setAccountChangedEventListener();
```

Comment:
I think this event listener part can be simplified to just

```
    const updateAccount = (network: string, dapiAccount: DapiAccount) =>
      this.dapi.getPublicKey().then(({ publicKey }) => {
        this.updateUserAccount({
          network,
          publicKey,
          ...dapiAccount,
        });
      }).catch((err) => {
        if (onError) {
          onError(err);
        }
      }); 

    let network: string | undefined;
    let account: DapiAccount | undefined;
    this.dapi.addEventListener(ACCOUNT_CHANGED, (dapiAccount) => {
      account = dapiAccount;
      if (network !== undefined) {
        updateAccount(network, dapiAccount);
      }
    });

    this.dapi.addEventListener(NETWORK_CHANGED, (dapiNetworks) => {
      network = dapiNetworks.defaultNetwork;
      this.updateNetworks(dapiNetworks);
      if (account !== undefined) {
        updateAccount(dapiNetworks.defaultNetwork, account);
      }
    });   
```

Reply:
Actually, the `getPublicKey` should also check if the address matches, i.e. it should look something like

```
    const updateAccount = (network: string, dapiAccount: DapiAccount) =>
      this.dapi.getPublicKey().then(({ address, publicKey }) => {
        if (address === dapiAccount.address) {
          this.updateUserAccount({
            network,
            publicKey,
            ...dapiAccount,
          });
        }
      }).catch((err) => {
        if (onError) {
          onError(err);
        }
      });   
```

Otherwise it's possible that `getPublicKey` returns a key for a different address from the account that's passed in.